src/Elastic.Apm/Metrics/MetricsProvider/SystemTotalCpuProvider.cs

```diff
@@ -0,0 +1,175 @@
+					return;
+
+				var total = numbers.Sum();
+				var idle = numbers[3];
+
+				_prevIdleTime = idle;
+				_prevTotalTime = total;
+			}
+		}
+
+		private bool _firstTotal = true;
+
+		private bool _getTotalProcessorTimeFailedLog;
+		private TimeSpan _lastCurrentTotalProcessCpuTime;
+		private DateTime _lastTotalTick;
+		private double _prevIdleTime;
+
+		private double _prevTotalTime;
+		private Version _processAssemblyVersion;
+		private PerformanceCounter _processorTimePerfCounter;
+
+		public int ConsecutiveNumberOfFailedReads { get; set; }
+		public string DbgName => "total system CPU time";
+
+		public IEnumerable<MetricSample> GetSamples()
+		{
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+			{
+				if (_processorTimePerfCounter == null)
+					_processorTimePerfCounter = new PerformanceCounter("Processor", "% Processor Time", "_Total");
+
+				var val = _processorTimePerfCounter.NextValue();
+
+				return new List<MetricSample> { new MetricSample(SystemCpuTotalPct, (double)val / 100) };
+			}
+
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+			{
+				if (File.Exists("/proc/stat"))
+				{
+					using (var sr = new StreamReader("/proc/stat"))
+					{
+						var firstLine = sr.ReadLine();
+						if (firstLine != null && firstLine.ToLower().StartsWith("cpu"))
+						{
+							var values = firstLine.Substring(5, firstLine.Length - 5).Split(' ');
+							if (values.Length < 4)
+								return null;
+
+							var numbers = new int[values.Length];
+
+							if (values.Where((t, i) => !int.TryParse(t, NumberStyles.Integer, CultureInfo.InvariantCulture, out numbers[i])).Any())
+								return null;
+
+							var total = numbers.Sum();
+							var idle = (double)numbers[3];
+
+							var idleTimeDelta = idle - _prevIdleTime;
+							var totalTimeDelta = total - _prevTotalTime;
+
+							var notIdle = 1.0 - idleTimeDelta / totalTimeDelta;
+
+							_prevIdleTime = idle;
+							_prevTotalTime = total;
+
+							return new List<MetricSample> { new MetricSample(SystemCpuTotalPct, notIdle) };
+						}
+					}
+				}
+			}
+
+			//The x-plat implementation is ~18x slower than perf. counters on Windows
+			//Therefore this is only a fallback in case of non-Windows OSs
+			//Also there are some issues with this code when e.g. the set of processes change between two runs.
+			var timeSpan = DateTime.UtcNow;
+			TimeSpan cpuUsage;
+
+			foreach (var proc in Process.GetProcesses())
+			{
+				try
+				{
+					cpuUsage += proc.TotalProcessorTime;
+				}
+				catch (Exception)
+				{
+					//can happen if the current process does not have access to `proc`.
+					if (_getTotalProcessorTimeFailedLog) continue;
+
+					_getTotalProcessorTimeFailedLog = true;
+					_logger.Info()
+						?.Log(
+							"The overall CPU usage reported by the agent may be inaccurate. This is because the agent was unable to access the CPU usage of some other processes");
+				}
+			}
+
+			var currentTimeStamp = DateTime.UtcNow;
+
+			if (!_firstTotal)
```

Comment:
For consistency it makes sense to have the first data read for X-plat fallback code in constructor as well, doesn't it?

Reply:
You were clearly not happy with this algorithm above, so I ended up with removing it (which I still don't think is a good decision, because this is a useful method for e.g. macOS users also with all its glitches). Therefore I think this is not an issue anymore.

Reply:
The only remaining issue that is a showstopper for me is mixing different sets of processes. With that issue fixed I am on board with having fallback algorithm that might have performance issues.
Fix with keeping PIDs is just a few more lines code but if any other fix will do as well.

Reply:
No, I gave up on this. Will be done on a follow up PR. I'd really like to make progress here and atm I see no other way than just giving up on things.